### PR TITLE
Support JSONP exports of conference schedules for embedding

### PR DESCRIPTION
**Pull request by kormoc**
Currently, it's rather difficult to use OSEM and embed a schedule
into another website. We currently use OSEM for the backend, but
still use a different system for our main site. This change
allows us to dynamically embed the schedule in our main website
as well as allow us a simple way to get data for the use in
mobile applications or similar.

This limits the JSONP support to only GETS on json formatted
requests. This prevents any XSS even though we dont require
a authenticity token.

### CODE REVIEW

File: app/controllers/conference_controller.rb [hunk 1]
```diff
@@ -1,6 +1,13 @@
 class ConferenceController < ApplicationController
   before_filter :respond_to_options
   load_and_authorize_resource find_by: :short_title
+  skip_before_filter :verify_authenticity_token, if: :json_request?
+
+  def json_request?
+    return false if request.request_method != 'GET'
+    return false if not request.format.json?
+    true
```
Comment (ex-nerd) [inline app/controllers/conference_controller.rb]: is it still ruby convention to leave off the `return`?  so weird...

File: app/serializers/conference_serializer.rb [hunk 2]
```diff
@@ -20,7 +20,9 @@ def revision
   # FIXME: adjusting the format the DIRTY way, for oSC13.
   # If you think this is ugly, don't look at the methods below
```
Comment (meonkeys) [inline app/serializers/conference_serializer.rb]: Heh.

**Comment by meonkeys** [comment]
:shipit:

Disclaimer: I've got no RoR skillz and don't know the OSEM codebase well.

File: app/controllers/conference_controller.rb [hunk 3]
```diff
@@ -1,6 +1,13 @@
 class ConferenceController < ApplicationController
   before_filter :respond_to_options
   load_and_authorize_resource find_by: :short_title
+  skip_before_filter :verify_authenticity_token, if: :json_request?
+
+  def json_request?
+    return false if request.request_method != 'GET'
+    return false if not request.format.json?
```
Comment (ChrisBr) [inline app/controllers/conference_controller.rb]:
```
unless request.format.json?
```

File: app/controllers/conference_controller.rb [hunk 4]
```diff
@@ -19,6 +26,11 @@ def schedule
     else
       @today = @conference.start_date.strftime('%Y-%m-%d')
     end
+
+    respond_to do |format|
+        format.html
+        format.json { render :json => @conference.as_json(), :callback => params[:callback] }
```
Comment (ChrisBr) [inline app/controllers/conference_controller.rb]: You should use the new Ruby hash syntax:

```
json: @conference.as_json()
```

File: app/models/conference.rb [hunk 5]
```diff
@@ -132,6 +132,73 @@ def date_range_string
   end
 
   ##
+  # Builds out the model for returning a json response
+  #
+  #
+  def as_json(options={})
+    super(
+        :only => [
```
Comment (ChrisBr) [inline app/models/conference.rb]: Here as well!

**Comment by ChrisBr** [comment]
@meonkeys thanks for your contribution :)

Hm, the only thing I worry is that we would publish events which aren't confirmed yet (but I'd be okay with that)... And I'm not sure if the as_json function is the right way to go ...

Additional, you have some style violations. You can run ```rubocop``` on your local machine to see the violations.

@hennevogel @differentreality what do you think?

**Comment by meonkeys** [comment]
@kormoc ping

**Comment by differentreality** [comment]
@ChrisBr Does it really show events that are not confirmed?
From what I am seeing it shows rooms, and then events that belong to those rooms. And since this is json format for schedule that makes sense. 
But in fact even if an event is confirmed, but not scheduled (assigned to room), it won't show.

Trying out the code locally alsoc concludes to this. Or I am missing sth :)

**Comment by differentreality** [comment]
@kormoc would you git rebase opensuse/master and then auto fix the syntax with 
**bundle exec rubocop -Dc .rubocop.yml -a**   please?

**Comment by kormoc** [comment]
Sorry for the branch churn. Rebasing mistype. The current PR should be good to go, with all the syntax fixes.

File: app/controllers/conference_controller.rb [hunk 6]
```diff
@@ -1,6 +1,13 @@
 class ConferenceController < ApplicationController
   before_filter :respond_to_options
   load_and_authorize_resource find_by: :short_title
+  skip_before_filter :verify_authenticity_token, if: :json_request?
```
Comment (hennevogel) [inline app/controllers/conference_controller.rb]: Why is this needed? verify_authenticity_token is for verification of the CSRF token for posting things, which you don't do.

Comment (differentreality) [inline app/controllers/conference_controller.rb]: http://api.rubyonrails.org/classes/ActionController/RequestForgeryProtection.html

Comment (hennevogel) [inline app/controllers/conference_controller.rb]: Ah there is an exception for ajax. Didn't know that. Please limit the skipping to the methods you really need it for. Not for the whole controller

File: app/models/conference.rb [hunk 7]
```diff
@@ -132,6 +132,73 @@ def date_range_string
   end
 
   ##
+  # Builds out the model for returning a json response
+  #
+  #
+  def as_json(_options={})
```
Comment (hennevogel) [inline app/models/conference.rb]: You do know about [serialization](http://api.rubyonrails.org/classes/ActiveModel/Serialization.html) right?

**Comment by hennevogel** [comment]
@kormoc Apart from the mentioned things I keep wondering what you are missing in the already existing API? You can query all the information you want.

**Comment by kormoc** [comment]
@hennevogel None of the current APIs I'm aware of allow jsonp formats. We currently embed the schedule on a different domain name and use client side javascript to get the updated schedule and to display it on the page. This allows us to use a fully static hosting solution for our main page but allow dynamic data on the sections that need it.

For example, http://seagl.org/schedule/2015.html is currently using this patch.

**Comment by differentreality** [comment]
@kormoc are you saying we could not integrate your changes into https://osem.seagl.org/api/v1/conferences instead of having something new?

**Comment by differentreality** [comment]
I talked with @kormoc on IRC, we can integrate his changes into our api instead.

But as seagl conf is coming up, he won't be able to work on it soon.

**Comment by differentreality** [comment]
Changes will be integrated in #729

**Comment by meonkeys** [comment]
:dancer:
